File: fx/src/main/java/gargoyle/fx/FXLauncher.java
package gargoyle.fx;

import gargoyle.fx.log.FXLog;
import javafx.application.Application;
import javafx.event.Event;
import javafx.stage.Stage;
import javafx.stage.WindowEvent;
import javafx.util.Callback;

import java.awt.Window;
import java.io.IOException;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.util.Locale;
import java.util.Map;
import java.util.MissingResourceException;
import java.util.ResourceBundle;
import java.util.logging.LogManager;

public final class FXLauncher {
    static final String KEY_APP = FXWrapper.class.getName();
    private static FXSplash splash;
    private static Window splashWindow;

    static {
        try {
            LogManager.getLogManager().readConfiguration(FXLauncher.class.getResourceAsStream("logging.properties"));
        } catch (final IOException e) {
            FXLog.error(e, FXUtil.format("{location} initialization error", Map.of("location", "log")));
        }
    }

    private FXLauncher() {
        throw new IllegalStateException(FXLauncher.class.getName());
    }

    public static void exit(final FXContext context) {
        final Stage primaryStage = context.getStage();
        Event.fireEvent(primaryStage, new WindowEvent(primaryStage, WindowEvent.WINDOW_CLOSE_REQUEST));
    }

    public static void notifySplash(final FXSplash.FXSplashNotification.Type type,
                                    final double progress, final String details) {
        notifySplash(new FXSplashNotificationImpl(type, progress, details));
    }

    public static void notifySplash(final FXSplash.FXSplashNotification splashNotification) {
        if (null != splash) {
            splash.handleSplashNotification(splashNotification);
        }
    }

    public static void requestPrevent(final FXContext context, final Callback<FXContext, FXCloseAction> callback) {
        FXCloseFlag.PREVENT.set(context, callback);
    }

    public static void requestRestart(final FXContext context) {
        FXCloseFlag.RESTART.set(context, fxContext -> {
            final FXCloseAction saved = FXCloseFlag.PREVENT.doIf(fxContext);
            try {
                final Stage stage = fxContext.getStage();
                final FXWrapper wrapper = getWrapper(fxContext);
                wrapper.askStop();
                wrapper.init();
                wrapper.start(stage);
            } catch (final Exception e) {
                FXLog.error(e, FXUtil.format("{location} initialization error",
                        Map.of("location", "close")));
                return FXCloseAction.CLOSE;
            }
            return FXCloseAction.KEEP == saved ? FXCloseAction.CLOSE : FXCloseAction.KEEP;
        });
        exit(context);
    }

    @SuppressWarnings("StaticMethodOnlyUsedInOneClass")
    public static void run(final Class<? extends FXApplication> appClass, final String[] args) {
        try {
            FXContextFactory.builder().setBaseClass(appClass).setLocale(Locale.getDefault());
            splashStart(appClass);
            final String[] newArgs = new String[args.length + 1];
            System.arraycopy(args, 0, newArgs, 1, args.length);
            newArgs[0] = appClass.getName();
            Application.launch(FXWrapper.class, newArgs);
        } catch (final Exception e) {
            FXLog.error(e, FXConstants.KEY_SPLASH);
            splashStop("");
        }
    }

    static void splashStop(final String details) {
        if (null != splashWindow) {
            if (splashWindow.isVisible()) {
                notifySplash(FXSplash.FXSplashNotification.Type.STOP, 1, details);
                splashWindow.dispose();
            }
        }
    }

    private static FXWrapper getWrapper(final FXContext fxContext) {
        return (FXWrapper) fxContext.getStage().getProperties().get(KEY_APP);
    }

    private static void splashStart(final Class<? extends FXApplication> appClass) {
        final String splashClassName = AccessController.doPrivileged((PrivilegedAction<String>) () ->
                System.getProperty(FXConstants.KEY_SPLASH_CLASS, FXConstants.SPLASH_CLASS_DEFAULT));
        try {
            final Class<? extends FXSplash> splashClass = FXIntUtil.loadClass(splashClassName, appClass.getClassLoader());
            splash = FXIntUtil.newInstance(splashClass);
            final String splashLocation = AccessController.doPrivileged((PrivilegedAction<String>) () ->
                    System.getProperty(FXConstants.KEY_SPLASH));
            splashWindow = splash.createWindow(null != splashLocation ? appClass.getResource(splashLocation)
                            : FXWrapper.class.getResource(FXConstants.SPLASH_DEFAULT),
                    ResourceBundle.getBundle(FXUtil.resolveBaseName(splashClass)));
            splashWindow.pack();
            splashWindow.setAlwaysOnTop(true);
            splashWindow.setLocationRelativeTo(null);
            splashWindow.setVisible(true);
            notifySplash(FXSplash.FXSplashNotification.Type.PRE_INIT, 0.0, "");
        } catch (final MissingResourceException | FXException e) {
            FXLog.warn(FXUtil.format("No resources {location}", Map.of("location", splashClassName)));
        }
    }

    public static final class FXSplashNotificationImpl implements FXSplash.FXSplashNotification {
        private final String details;
        private final double progress;
        private final Type type;

        private FXSplashNotificationImpl(final Type type, final double progress, final String details) {
            this.type = type;
            this.progress = progress;
            this.details = details;
        }

        @Override
        public String getDetails() {
            return details;
        }

        @Override
        public double getProgress() {
            return progress;
        }

        @Override
        public Type getType() {
            return type;
        }
    }
}

File: fx/src/main/kotlin/gargoyle/fx/FxLauncher.kt
package gargoyle.fx

import gargoyle.fx.FxSplash.FxSplashNotification
import gargoyle.fx.log.FxLog
import gargoyle.fx.log.FxLoggerFactory
import gargoyle.fx.log.jul.FxJULLogger
import javafx.application.Application
import javafx.event.Event
import javafx.stage.WindowEvent
import java.awt.Window
import java.util.Locale
import java.util.MissingResourceException
import java.util.ResourceBundle
import kotlin.reflect.KClass

@Suppress("MemberVisibilityCanBePrivate")
object FxLauncher {

    class FxSplashNotificationImpl(
        override val type: FxSplashNotification.Type,
        override val progress: Double,
        override val details: String
    ) : FxSplashNotification

    val KEY_APP: String = FxWrapper::class.qualifiedName!!
    private lateinit var splash: FxSplash
    private lateinit var splashWindow: Window

    init {
        FxLoggerFactory.configure(FxJULLogger, FxJULLogger::class.java.getResource("logging.properties"))
    }


    fun exit(context: FxContext): Unit =
        Event.fireEvent(context.stage, WindowEvent(context.stage, WindowEvent.WINDOW_CLOSE_REQUEST))

    fun notifySplash(type: FxSplashNotification.Type, progress: Double, details: String): Unit =
        notifySplash(FxSplashNotificationImpl(type, progress, details))

    fun notifySplash(splashNotification: FxSplashNotification) {
        splash.handleSplashNotification(splashNotification)
    }

    fun requestPrevent(context: FxContext, callback: (FxContext) -> FxCloseAction) {
        FxCloseFlag.PREVENT[context] = callback
    }

    fun requestRestart(context: FxContext) {
        FxCloseFlag.RESTART[context] = { it: FxContext ->
            with(it) {
                FxCloseFlag.PREVENT.doIf(it)
                    .let { closeAction ->
                        try {
                            (this.stage.properties[KEY_APP] as FxWrapper?)?.let {
                                with(it) {
                                    askStop()
                                    init()
                                    start(stage)
                                }
                            }
                            when (closeAction) {
                                FxCloseAction.KEEP -> FxCloseAction.CLOSE
                                else -> FxCloseAction.KEEP
                            }
                        } catch (e: Exception) {
                            FxLog.error("{location} initialization error", e)
                            FxCloseAction.CLOSE
                        }
                    }

            }
        }
        exit(context)
    }

    fun run(appClass: KClass<out FxApplication>, args: Array<String>): Unit =
        try {
            FxContext.contextBuilder = FxContext.builder().setBaseClass(appClass).setLocale(Locale.getDefault())
            splashStart(appClass)
            Application.launch(FxWrapper::class.java, *arrayOf(appClass.qualifiedName!!) + args)
        } catch (e: Exception) {
            FxLog.error(FxConstants.KEY_SPLASH, e)
            splashStop("")
        }

    fun splashStop(details: String) {
        if (::splashWindow.isInitialized) {
            splashWindow.let {
                if (it.isVisible) {
                    notifySplash(FxSplashNotification.Type.STOP, 1.0, details)
                    it.dispose()
                }
            }
        }
    }

    private fun splashStart(appClass: KClass<out FxApplication>) {
        val splashClassName = FxIntUtil.doPrivileged<String> {
            System.getProperty(FxConstants.KEY_SPLASH_CLASS, FxConstants.SPLASH_CLASS_DEFAULT)
        }
        try {
            val splashClass: KClass<out FxSplash> = FxIntUtil.loadClass(splashClassName, appClass.java.classLoader)
            splash = FxIntUtil.newInstance(splashClass)
                .also { fxSplash ->
                    with(
                        fxSplash.createWindow(FxIntUtil.doPrivileged<String?> { System.getProperty(FxConstants.KEY_SPLASH) }
                            ?.let { appClass.java.getResource(it) }
                            ?: FxWrapper::class.java.getResource(FxConstants.SPLASH_DEFAULT),
                            ResourceBundle.getBundle(
                                splashClass.qualifiedName!!.replace('.', '/'),
                                Locale.getDefault(),
                                splashClass.java.classLoader
                            )
                        )
                    ) {
                        pack()
                        isAlwaysOnTop = true
                        setLocationRelativeTo(null)
                        isVisible = true
                        splashWindow = this
                    }
                }
            splash.also { notifySplash(FxSplashNotification.Type.PRE_INIT, 0.0, "") }
        } catch (e: MissingResourceException) {
            FxLog.warn("No resources $splashClassName")
        } catch (e: FxException) {
            FxLog.warn("Other error from $splashClassName")
        }
    }

}
